internal/venice-common/src/main/java/com/linkedin/venice/blobtransfer/NettyP2PBlobTransferManager.java

```diff
@@ -37,25 +41,36 @@ public void start() throws Exception {
   @Override
   public CompletionStage<InputStream> get(String storeName, int version, int partition)
       throws VenicePeersNotFoundException {
-    CompletionStage<InputStream> inputStream;
+    CompletableFuture<InputStream> inputStreamFuture = new CompletableFuture<>();
     BlobPeersDiscoveryResponse response = peerFinder.discoverBlobPeers(storeName, version, partition);
+
     if (response == null || response.isError()) {
       throw new VenicePeersNotFoundException("Failed to obtain the peers for the requested blob");
     }
+
     List<String> discoverPeers = response.getDiscoveryResult();
     if (discoverPeers == null || discoverPeers.isEmpty()) {
       throw new VenicePeersNotFoundException("No peers found for the requested blob");
     }
-    try {
-      // TODO: add some retry logic or strategy to choose the peers differently in case of failure
-      // instanceName comes as a format of <hostName>_<applicationPort>
-      String chosenHost = discoverPeers.get(0).split("_")[0];
-      inputStream = nettyClient.get(chosenHost, storeName, version, partition);
-    } catch (Exception e) {
-      throw new VenicePeersNotFoundException("The connection to peers failed", e);
+
+    for (String peer: discoverPeers) {
+      try {
+        // instanceName comes as a format of <hostName>_<applicationPort>
+        String chosenHost = peer.split("_")[0];
+        CompletionStage<InputStream> result = nettyClient.get(chosenHost, storeName, version, partition);
+        InputStream inputStream = result.toCompletableFuture().get();
```

Comment:
There is no point to wrap original inputStream with different layer of futures... Just catch the exception here, if error is thrown that means connection is broken, the node might just not exist. For other types of errors I think we can deal with them in the follow up PR.
And you are returning a different type a future, which is different from the method signature, IDK if that even compiles.

Can you just find the first host that does not fail the connection and return it? (Plus print the chosen host as well?)

Reply:
Did i do what you meant in this new push?